### ISSUE TITLE
Yet another API attempt

### ISSUE THREAD

**Issue by githubshrek** [Tatoeba/tatoeba2#2669]
Author: Hugh
Tatoeba: https://tatoeba.org/eng/user/profile/tatoebashrek 

# Introduction
## Problem definition
I would like to build an API for Tatoeba to allow easy and direct use by other websites.

I am the owner of the language-learning website https://lingopolo.org/ which is a sort of cross between Anki (flashcard tool), Duolingo (language-learning tool) and Tatoeba. It builds language lessons automatically based on the added sentence content. 

Recently we have started to use content from Tatoeba, e.g. see the Tatoeba acknowledgement at the bottom of these pages:
* https://lingopolo.org/thai/word/why-didnt-you-call-me-last-week 
* https://lingopolo.org/thai/word/i-ran-across-my-old-teacher-on-the-street 
* https://lingopolo.org/dutch/word/tom-hasnt-eaten-enough 
* https://lingopolo.org/dutch/word/most-of-my-friends-are-still-single 

What we would like to do is automatically add sentences from Tatoeba, directly and programatically from the Lingopolo website. The Lingopolo website would regularly, via cron, make requests to Tatoeba of the type:
* give me the ids of 100 sentences in NL with an EN translation and audio
* for each of the ids, give me the full content of that particular sentence, such as:
  * id
  * EN text
  * NL text
  * copyright information
  * user
  * audio

## Proposed solution summary
This describes part of a possible Tatoeba API, which would allow users to directly via API (on a URL):
query the ids of sentences matching a particular tag
query an individual sentence

I see that there have been some attempts at building an API before:
* in 2014, https://github.com/Akirato/tatoeba-api 
* in 2015, https://github.com/Tatoeba/tatoeba-api 
* in 2020, https://github.com/Tatoeba/tatoeba2/pull/2341 

## Lingopolo
Lingopolo consists of 2 developers, myself and Alex. It is based on Drupal (which means it is written in PHP like Tatoeba). We would anticipate that we would do most of the development of this API.

## Useful links on API structure
There are many external APIs which it would be worth studying to get the best structure, but here are a few recent links I was looking at.

Wikidata API:
* https://www.wikidata.org/wiki/Wikidata:Data_access 

See how Wikidata offers two formats, one which shows HTML and one which shows Json, by adding the "format=json" parameter:
* https://www.wikidata.org/w/api.php?action=wbgetentities&ids=Q9128 
* https://www.wikidata.org/w/api.php?action=wbgetentities&ids=Q9128&format=json 

In the description below I use 2 different formats inconsistently for the API:
* /api.php?action=show_sentences_with_tag=7770&offset=0&limit=100
* GET /sentence/1111401

# Get the list of sentences which match a given tag
## HTML
https://tatoeba.org/eng/tags/show_sentences_with_tag/7770 

## Example API call
/api.php?action=show_sentences_with_tag=7770&offset=0&limit=100

Pagination based on https://nordicapis.com/everything-you-need-to-know-about-api-pagination/  

## Example Response
100 rows in the following format:
```
{
    "id": 6549421,
    "id": 6627714,
    "id": 6996983,
    "id": 2722194,
    …
}
```

# Get a sentence
This call returns the detail of a single sentence.

## Example: "Tom asked me who walked home with me last night."
https://tatoeba.org/eng/sentences/show/6432721 

```
{
    "id": 6432721,
    "text": "Tom asked me who walked home with me last night.",
    "language": "en",
    "derivedFrom": "",
    "belongsTo": "CK",
    "tags": [],
    "textLicence": "CC BY 2.0 FR",
    "audioLicence": "CC BY 2.0 FR",
    "audio": "https://audio.tatoeba.org/sentences/eng/6432721.mp3",
    "translations": [
        {
            "id": 9672097,
	"language": "epo",
	"text": "Tomo demandis al mi, kiu marŝis hejmen kun mi hieraŭ nokte."
        }
     ]
}  
``` 

## Example "VK is de afkorting van het Verenigd Koninkrijk"
Gets the information for a single sentence such as https://tatoeba.org/eng/sentences/show/1111401 

This example demonstrates how the direct and indirect translations would be arranged in two separate arrays.

Example call:
GET /sentence/1111401

Response:
```
{
    "id": 1111401,
    "lang": "nl",
    "text": "VK is de afkorting van het Verenigd Koninkrijk",
    "textLicence": "CC BY 2.0 FR",
    "audioLicence": "CC BY 2.0 FR",
    "owner": "megamanenm",
    "directTranslations": [
        {
            "id": 1067763,
            "lang": "en",
            "text": "UK is the abbreviation for the United Kingdom"
        }
        {
            "id": 1208155,
            "lang": "esp",
            "text": "UR estas mallongigo por Unuiĝinta Reĝlando."
        }
    ]
    "indirectTranslations": [
        {
            "id": 7527207
 	"lang": "ber",
            "text": "TY d asegzel n Tgelda Yedduklen."
        }
    ]
}
```

# Get user (member)
Get the details of a particular user (member).

## HTML
https://tatoeba.org/eng/user/profile/Guybrush88

## Example API call
/api.php?action=get_user&user=Guybrush88

## Example response

```
{
    "user": "Guybrush88",
    "country": "Italy",
    "birthday": "1988-12-27",
    "homepage": "",
    "commentsPosted": 6717,
    "sentencesOwned": 687581,
    "audioRecordings": 0,
    "sentencesFavorited": 45,
    "contributions": 2243651,
    "languages": [
        {
        "language": "it",
        "level": 5,
        "details": "Native speaker"
        }
        {
        "language": "en",
        "level": 4,
        "details": "C1 level of Cambridge, C2 of my University"
        }
        {
        "language": "fr",
        "level": 4,
        "details": "C1 level of my University"
        }
        {
        "language": "egl",
        "level": 4,
        "details": "The dialect from my province"
        }
}
```      

# Get the list of supported languages

This is not seen as currently required, but is included as one of the many other API endpoints which might eventually be built.

returns something like:

```      
{
     "languages": ["en", "fr", "nl", ....]
}
```

**Comment by jiru** [comment]
Thank you so much for sharing your use case with us. We are getting people asking for an API every now and then.

The contents of the json returned by the API look good. Except perhaps the "licence" key. I think we need to make it clearer that there is a license for the text and a license for the audio, by including both of them upfront in the json.

> The Lingopolo website would regularly, via cron, make requests to Tatoeba of the type:
> * give me the ids of 100 sentences in NL with an EN translation and audio

Could you elaborate on this? What do you do with the sentences, i.e. what’s the complete workflow? Do you want to get new ids at random, or are you trying to get new sentence pairs as they are being added on Tatoeba? When getting the results, how to make sure that you haven’t got some of them already from a previous call?

Additional question: what would be the best way to handle the removal, the edition or the unlink of sentences you previously downloaded using the API?

**Comment by githubshrek** [comment]
Thanks for the feedback. I didn't realise there were 2 licences. 

# Licences

I have to say, it's not 100% clear what the first "License" is referring to:
![image](https://user-images.githubusercontent.com/7943077/112892860-29978c00-90da-11eb-902b-738e702d0b53.png)

[As an aside, I would suggest to make these 2 headings say "Text" and "Audio" for consistency].

Guesswork and reading https://en.wiki.tatoeba.org/articles/show/faq makes me think that the following would describe things correctly:
    "textLicence": "CC BY 2.0 FR",
    "audioLicence": "CC BY 2.0 FR",

I have changed the original post to use this new format.

# Workflow

I have not yet worked out all the details of the Lingopolo end, but I know that if I have 2 API calls available which do:

* give me the sentence data for a given id
* give me a list of Tatoeba sentence ids which match whatever search criteria I wish (basically the Advance Search https://tatoeba.org/eng/sentences/advanced_search via API), e.g.:
  * tag
  * source language
  * target language
  * has audio
  * is approved
  * etc.

then with these 2 API calls I can do everything I would ever want from the Lingopolo end.

## Automatic input of all the relevant Tatoeba fields for a single sentence

Let me unpack now vs. future for a single sentence.

**Currently**, this is what a Lingopolo editor has to do, to go from a sentence in Tatoeba to having that same sentence correctly in Lingopolo:
* (firstly find a suitable sentence)
* create a new Lingopolo entry
* copy the EN text into the Lingopolo entry, into the EN field
* copy the non-EN text into the non-EN field
* copy the Tatoeba licence into the EN text licence field
* (copy the Tatoeba licence into the non-EN text licence field) - this step is currently ignored, and only the EN licence is filled, but this is strictly incorrect, since Lingopolo is using the Tatoeba translation, so that field should also be filled
* download the audio (if exists) from Tatoeba entry
* upload the audio file into the Lingopolo entry
* fill in the audio licence field

The first step would be automating all of this:
* editor on Lingopolo adds the Tatoeba URL (or the sentence id)
* editor clicks on "Get Tatoeba data" 

and the API would automatically extract all of the data from Tatoeba, and Lingopolo would automatically fill in all of the fields correctly with the correct data.

## Get multiple ids
The next step would be to do this for a bunch of ids, collected from a search, e.g. 100 which match a given criteria selected by the editor.

## Get via cron
The final step would be to get this all done via cron. 

# Answers to your questions
Q. Do you want to get new ids at random, or are you trying to get new sentence pairs as they are being added on Tatoeba? 
A. Random is probably fine to begin with. The only thing I have noticed (to my disappointment) is that the quality of some of the translations done by non-native speakers, which my non-native English editors copied and used, not realising that the English was wrong and I had to completely delete. In future I will definitely restrict selection to:
* "List 907" tag
* "is approved"
* pre-checked lists like http://www.manythings.org/bilingual/

Q. When getting the results, how to make sure that you haven’t got some of them already from a previous call?
There are 2 ways I would do this:
* check that the EN and non-EN are unique
* check the source field to see if it is a Tatoeba link and compare the sentence number

Q. Additional question: what would be the best way to handle the removal, the edition or the unlink of sentences you previously downloaded using the API?
A. I think that's an issue that's 5 years away at least. I suppose you mean if Tatoeba deletes or removes a sentence. In that case, I would probably have some cleanup job that checks the Tatoeba links are still valid

**Comment by jiru** [comment]
Thank you very much! I now have a good overview of your use case. Search + sentence retrieval is probably the most wanted API feature so it is definitely a good start. I wonder if we want to expose all the advanced search criteria because some criterion like "owner" doesn’t make so much sense outside of Tatoeba. As a course editor, I imagine that you would need to go back and forth between Tatoeba and your own API-connected interface, just to check what owner, tag or list to filter with. This makes me wonder if this approach is such a good way to go for your use case after all. This is just an idea, but since we already provide searching at Tatoeba, what if the course editor would search on tatoeba.org instead, add all the needed sentences to a list (a list is a feature of Tatoeba), and then import that list into the course using a single API call?

By the way your story reminds me that the author of [LibreLingo](https://github.com/kantord/LibreLingo) who contacted us recently with what looked like a similar use case.

**Comment by mramosch** [comment]
I would really need e.g. an endpoint that returns a

•  [sentence language] to [sentence owner skill level in that language] correlation

i.e. submitting a sentence ID and getting back a language skill level (0, 1, 2, 3, 4, 5 or unspecified)

**Comment by githubshrek** [comment]
> I would really like to see e.g. an endpoint that returns a
> 
> • [sentence language] to [sentence-owner’s skill level in that language] correlation
> 
> i.e. submitting a sentence ID and getting back a language skill level (0, 1, 2, 3, 4, 5 or unspecified)
> 
> Or at least a [sentence ID] belongs to [User ID] correlation...
> 
> I don’t wanna hit the server with a complete page request for every direct and indirect translation of a sentence and parse all the responses for the usernames of the owner of the translations just in order to circumvent the download of several hundred gigaBytes of download files...

I understand where you're coming from. I too am not going to be interested in pulling down Italian sentences owned by a Japanese, or English sentences owned by a Korean etc. so making sure that the owner is a native speaker of the language is clearly an important criteria.

Nonetheless, for a first step, I think it's important to try and keep the API very simple, allowing you to fetch the details of specific entities, in particular a sentence, but it certainly makes sense to also be able to query a particular user (member). I have added to the original description above a "**Get user (member)**" API call, which would give all the details about a particular user. Since the sentence has a "belongsTo" key then you would be able to go from sentence to get all the details about the sentence owner.

**Comment by githubshrek** [comment]
OK, I think I've written the draft version of the one part of the requirement, which was:

for each of the ids, give me the full content of that particular sentence, such as:
- id
- EN text
- NL text
- copyright information
- user
- audio

The way to use will be by putting /json/ in the URL instead of /show/ for the sentence, e.g. for sentence 3

* http://localhost:8080/eng/sentences/json/3 show the Json view
* http://localhost:8080/eng/sentences/show/3 show the normal view

Displayed would be the sentence with all translations, e.g.
![image](https://user-images.githubusercontent.com/7943077/116626398-707fd800-a94b-11eb-953f-478cb30c42cd.png)

It seems I only needed to make a few small changes in 2 places:
(1) add 'json' as a URL option for the sentences in auth_actions.php
![image](https://user-images.githubusercontent.com/7943077/116626520-a6bd5780-a94b-11eb-9e27-081fbf85823c.png)

(2) Write a little json() method to simply dump out the Sentence object:
![image](https://user-images.githubusercontent.com/7943077/116626621-d8ceb980-a94b-11eb-8ae9-86e04b4d299d.png)

If this looks OK to folk here, I'll look at how to do the pull request...

**Comment by jiru** [comment]
> If this looks OK to folk here, I'll look at how to do the pull request...

I would be happy to review and merge your code if you create a PR. However, it is not in a state that we can make it the "official" API yet. It would be more like an "unofficial" API only for "insiders" until we can come up with something better. I have been thinking about what’s missing before we can reach this "public" state. First of all, I think the URLs are not good:

> The way to use will be by putting /json/ in the URL instead of /show/ for the sentence, e.g. for sentence 3
> 
>     * http://localhost:8080/eng/sentences/json/3 show the Json view
> 
>     * http://localhost:8080/eng/sentences/show/3 show the normal view

It looks like you tried to make your API match the website URLs, but I don’t think it’s a good design. Website URLs are not stable and we are free to change them as long as it looks good in everyone’s browser. The URLs of the API, in contrast, should be stable. So better move them to an entire different name space such as a /api prefix (following your naming, that would be https://tatoeba.org/api/sentences/json/3). However, `api` happens to be a valid language code for Apiaká, so it could create a conflict or prove otherwise confusing. So maybe https://api.tatoeba.org/sentences/json/3 would be better.

I don’t think it is relevant to include "json" in the URL if all the API uses json, so we can shorten it to https://api.tatoeba.org/sentences/3.

Finally, maybe we should include versioning in the URL in case we want to change the API at some point (I am not sure what are the best practices): https://api.tatoeba.org/v1/sentences/3.

All that should be possible to implement using [CakePHP routing](https://book.cakephp.org/3/en/development/routing.html). For all these reasons, your API code would be better moved into a different controller, for instance ApiController.php.

And then we need to add unit tests for the API.

**Comment by githubshrek** [comment]
I had not read your comment above when I started to implement the "search" part of the API. For that, there wasn't an easy equivalent to the normal /search URL, so I began with /search_json which I rather hated, so I changed to:

- /api/search
- /api/sentence

![image](https://user-images.githubusercontent.com/7943077/116827882-cb0f7300-ab9b-11eb-9cb2-285f8c2c4af6.png)

I think this URL schema is better.

Regarding versioning, yes, I think it is best practice to include a version number. However, since we currently only have one version, I think it's OK too to say that the current URLs will always give the most recent version of the API. You can imagine a future where we have something like:

- /api/search (which now delivers v2)
- /api/v1/search
- /api/v2/search

Regarding tests, yes, no doubt you're right. I haven't included any yet.

**Comment by githubshrek** [comment]
For the record, here is sample output from the two API calls:

The **sentence** detail at /api/sentence/[sentence-id]
![image](https://user-images.githubusercontent.com/7943077/116828124-08283500-ab9d-11eb-88b1-0477057655eb.png)

The **search** at /api/search?from=eng&has_audio=&native=&orphans=no&query=&sort=relevance&sort_reverse=&tags=&to=fra&trans_filter=limit&trans_has_audio=&trans_link=&trans_orphan=&trans_to=&trans_unapproved=&trans_user=&unapproved=no&user=

which has the identical format to the normal search

http://localhost:8080/eng/sentences/search?from=eng&has_audio=&native=&orphans=no&query=&sort=relevance&sort_reverse=&tags=&to=fra&trans_filter=limit&trans_has_audio=&trans_link=&trans_orphan=&trans_to=&trans_unapproved=&trans_user=&unapproved=no&user=

and will just return a list of sentence ids:
![image](https://user-images.githubusercontent.com/7943077/116828186-67864500-ab9d-11eb-9c79-3b25ce8de143.png)

**Comment by jiru** [comment]
I am glad you are making progress on this. :slightly_smiling_face: 

> Regarding versioning, yes, I think it is best practice to include a version number. However, since we currently only have one version, I think it's OK too to say that the current URLs will always give the most recent version of the API. You can imagine a future where we have something like:
>
> - /api/search (which now delivers v2)
> - /api/v1/search
> - /api/v2/search

I get your point but I don’t think this is a good idea. Version updates are here to indicate a breaking change. With such setup, if API users use the versionless endpoint /api/search and we update the API to a new version, it will certainly and suddenly break their client. That’s why I think it is better to *enforce* specifying the version number and making it visible.

**Comment by githubshrek** [comment]
OK, I'll change to /api/v1/ 

Here is some reading on the matter:
- https://www.freecodecamp.org/news/how-to-version-a-rest-api/
- https://www.xmatters.com/blog/devops/blog-four-rest-api-versioning-strategies/
- https://restfulapi.net/versioning/

**Comment by githubshrek** [comment]
I would like to test the search on a large number of sentences (currently I have created 3 in my imouto version). 

https://github.com/Tatoeba/tatoeba2/blob/dev/README.tatovm.md says:
"The default install contains no sentences at all and the search won’t work. You will have to add a few sentences on your own."

That's fine for testing with a few sentences, but is there an easy way to import a set? I don't need millions, but I would like to have at least 100 or so to make sure there are no paging issues (e.g. the Json search array being cut off after 10 results).

**Comment by jiru** [comment]
Unfortunately, you’re on your own for this. There is an open issue: #2010. I am ashamed to admit that I often just run my code on dev.tatoeba.org since I have access. But if it is just a matter of having more than 10 sentences, it’s probably quicker to just add them manually.

**Comment by githubshrek** [comment]
I added a dozen sentences, and true enough the paging was kicking in and limiting my API results to 10 at a time. I've changed the value to a hard-coded 100 results. It's not not beautiful or ideal, but I think it's OK for a first version.

One thing I noticed is that the way controllers and URL paths work in auth_actions.php is that all URLs are structurally only 2 deep, e.g. /sentences/show ;

- first level implies the controller (e.g. 'sentences' uses the SentencesController.php)
- second level implies the method in the controller (e.g. show() )

This means that a structure like /api/v1/show does not look like it is easily handled. I have therefore chosen the ugly /api_v1/show structure.

**Comment by jiru** [comment]
I think we can use custom routes to handle this. I suggest you submit a PR with /api_v1/show and I will take care of making the URL pretty later.

**Comment by alanfgh** [comment]
Although this is not the main issue addressed on the page, I want to respond to this comment:

> In future I will definitely restrict selection to subsets like:
>    "List 907" tag
>    "is approved"
>   pre-checked lists like http://www.manythings.org/bilingual/

There is no "is approved". There is an "Is unapproved" criterion, which means that a sentence was marked as untrustworthy, probably because it was added by an author who is known to have contributed material that cannot be used at Tatoeba because it is improperly taken from elsewhere. Sentences that are marked "Is unapproved" should definitely be excluded, but that's a very low bar. It does not say anything about the quality of the sentence.

The first and third items, "List 907" and pre-checked lists from manythings.org, are essentially the same thing, since the person who runs manythings.org is also the owner of List 907. The major problem here is that restricting sentences to that list eliminates a great deal of diversity, since the owner tends to favor sentences that:
- he has written himself
- are short
- are aimed at beginning language learners
- are often near-duplicates of others on the list ("I think that we should tell Tom that we won't do that anymore." "I think we should tell Tom we won't do that anymore..")

There are also sentences within the list that contain errors,

By contrast, restricting sentences to those contributed by self-declared native-level speakers is a better way to increase diversity, and cover a broader range of vocabulary and sentence structure, though you might also need to review them more closely.

**Comment by githubshrek** [comment]
For my own documentation as much as for anybody else, the API calls are _currently_ as the following examples:

* http://localhost:8080/eng/api_v1/sentence/1 
* http://localhost:8080/eng/api_v1/search?from=eng&has_audio=&native=&orphans=no&query=&sort=relevance&sort_reverse=&tags=&to=fra&trans_filter=limit&trans_has_audio=&trans_link=&trans_orphan=&trans_to=&trans_unapproved=&trans_user=&unapproved=no&user=

**Comment by trang** [comment]
@githubshrek I've merged your PR and deployed the changes on our dev website. You can try it out:

- https://dev.tatoeba.org/eng/api_v0/sentence/1
- https://dev.tatoeba.org/eng/api_v0/search?from=eng&has_audio=&native=&orphans=no&query=&sort=relevance&sort_reverse=&tags=&to=fra&trans_filter=limit&trans_has_audio=&trans_link=&trans_orphan=&trans_to=&trans_unapproved=&trans_user=&unapproved=no&user=

**Comment by ckjpn** [comment]
### My experiments

Search: Canadian
from: eng
to: jpn
list: 907
sort: created
direct translations only

https://dev.tatoeba.org/eng/api_v0/search?from=eng&list=907&query=Canadian&to=jpn&trans_filter=limit&trans_link=direct&trans_to=jpn


Search: Boston
from: eng
to: jpn
audio: yes
sort: created
(direct and indirect translations)

https://dev.tatoeba.org/eng/api_v0/search?from=eng&trans_filter=limit&query=Boston&sort=created&has_audio=yes&trans_to=jpn

**Comment by githubshrek** [comment]
I did some comparisons, e.g.:

- https://dev.tatoeba.org/en/sentences/search?from=vol&query=dog+bunom&to=
- https://dev.tatoeba.org/en/api_v0/search?from=vol&query=dog+bunom&to= (using https://jsonformatter.org/json-pretty-print to beautify)

and

- https://dev.tatoeba.org/en/sentences/show/1011293
- https://dev.tatoeba.org/en/api_v0/sentence/1011293

Looks correct and good.

I like how 100% the normal search matches the API search, which makes it super easy to experiment with the results.

**Comment by ckjpn** [comment]
### Suggestions

These are only meant as food for thought, and not as a criticism of what's been done.


1.

Looking at the last line of the output from my "Boston" search above, using "jason-pretty print" I noticed that the output also includes lists that owners have chosen to have "unlisted" (or perhaps "private").  Maybe the wishes of list owners should be considered, even though this data is in the exported files.

https://tatoeba.org/en/sentences_lists/index/For%20Ori
4088	tommy_san	2015-01-28 07:26:52	2015-05-17 00:09:45	For Ori	creator


```
  {
    "id": 2990191,
    "text": "I'd like to visit Boston at least one more time before I die.",
    "lang": "eng",
    "correctness": 0,
    "script": null,
    "license": "CC BY 2.0 FR",
    "translations": [
      [
        {
          "id": 2990547,
          "text": "死ぬまでにあと1回はボストンに行きたい。",
          "lang": "jpn",
          "correctness": 0,
          "script": null,
          "transcriptions": [
            {
              "id": 1728492,
              "sentence_id": 2990547,
              "script": "Hrkt",
              "text": "[死|し]ぬまでにあと[1|][回|かい]はボストンに[行|い]きたい。",
              "user_id": null,
              "needsReview": true,
              "modified": "2020-02-11T06:34:35+00:00",
              "user": null,
              "readonly": false,
              "type": "altscript",
              "html": "<ruby>死<rp>（</rp><rt>し</rt><rp>）</rp></ruby>ぬまでにあと<ruby>1<rp>（</rp><rt></rt><rp>）</rp></ruby><ruby>回<rp>（</rp><rt>かい</rt><rp>）</rp></ruby>はボストンに<ruby>行<rp>（</rp><rt>い</rt><rp>）</rp></ruby>きたい。",
              "markup": "死｛し｝ぬまでにあと1｛｝回｛かい｝はボストンに行｛い｝きたい。",
              "info_message": "This alternative script may contain errors because it has been generated by a piece of software that is not perfect."
            }
          ],
          "audios": [],
          "isDirect": true,
          "lang_name": "Japanese",
          "dir": "ltr",
          "lang_tag": "ja"
        },
        {
          "id": 2991900,
          "text": "死ぬまでにあと1回ボストンに行きたい。",
          "lang": "jpn",
          "correctness": 0,
          "script": null,
          "transcriptions": [
            {
              "id": 1728500,
              "sentence_id": 2991900,
              "script": "Hrkt",
              "text": "[死|し]ぬまでにあと[1|][回|かい]ボストンに[行|い]きたい。",
              "user_id": null,
              "needsReview": true,
              "modified": "2020-02-11T06:34:35+00:00",
              "user": null,
              "readonly": false,
              "type": "altscript",
              "html": "<ruby>死<rp>（</rp><rt>し</rt><rp>）</rp></ruby>ぬまでにあと<ruby>1<rp>（</rp><rt></rt><rp>）</rp></ruby><ruby>回<rp>（</rp><rt>かい</rt><rp>）</rp></ruby>ボストンに<ruby>行<rp>（</rp><rt>い</rt><rp>）</rp></ruby>きたい。",
              "markup": "死｛し｝ぬまでにあと1｛｝回｛かい｝ボストンに行｛い｝きたい。",
              "info_message": "This alternative script may contain errors because it has been generated by a piece of software that is not perfect."
            }
          ],
          "audios": [],
          "isDirect": true,
          "lang_name": "Japanese",
          "dir": "ltr",
          "lang_tag": "ja"
        }
      ],
      []
    ],
    "users_sentences": [
      {
        "sentence_id": 2990191,
        "correctness": 1
      }
    ],
    "sentences_lists": [
      {
        "id": 3725
      },
      {
        "id": 3726
      },
      {
        "id": 3729
      },
      {
        "id": 3730
      },
      {
        "id": 3731
      },
      {
        "id": 3732
      },
      {
        "id": 3733
      },
      {
        "id": 3735
      },
      {
        "id": 3737
      },
      {
        "id": 4088
      },
      {
        "id": 907
      },
      {
        "id": 4000
      },
      {
        "id": 7389
      },
      {
        "id": 7412
      }
    ],
    "transcriptions": [],
    "audios": [
      {
        "author": "CK"
      }
    ],
    "user": {
      "username": "CK"
    },
    "lang_name": "English",
    "dir": "ltr",
    "lang_tag": "en",
    "is_favorite": false,
    "is_owned_by_current_user": true,
    "permissions": {
      "canEdit": true,
      "canTranscribe": false,
      "canReview": true,
      "canAdopt": true,
      "canDelete": true,
      "canLink": true
    },
    "current_user_review": 1
  }
```



2. Another thing to consider, perhaps, is an alternative "light" version of this that would include less data.  This could possibly reduce the load on tatoeba.org and might possibly mean that such requests would be handled more quickly.

3. I'd love to see the owners' names for the translations, too.  This might help make it possible to only show translations by certain trusted members (native speakers, or by some other method), and might make it possible to filter-out translations by those the user of the data doesn't trust.

**Comment by githubshrek** [comment]
@ckjpn 

Just a bit of context to start. When I set out to build the API, I imagined I would manually have to pull each of the different elements together and literally build the Json by hand. In fact, the super good news is that the API simply calls the existing code and returns the results as Json format; basically a 1-line piece of code, with almost no programming effort. I say this because the downside is that it is not so easy at all to change what is returned, since we just display what we get. Sure, it would be possible to customise things, either by changing the code that provides the data or by tweeking the return values. In either case it would be probably at least 10 times as much work for very little gain.

Now concretely addressing each of the points you raised:

### 1. sentences_lists

Since I wasn't building the Json myself, I didn't even know it was adding these list numbers. 

Since the list ids are just numbers, they don't give any real information. If you have a list number like 8432 then sure you can go to URL https://dev.tatoeba.org/en/sentences_lists/show/8432 and see what is there, but if the list is marked as private and it displays then that is a problem with the privacy management.

### 2. consider, perhaps, is an alternative "light" version

Originally the search API was written to only return a list of sentence ids. As @trang pointed out, there is not really a use case where you would then need to ask Tatoeba for the detail on each of the sentences. Therefore, the search API directly returns all the details. Notice that the search is a paged request, meaning that if there are thousands of results, the query will only return one page of results at a time. To be honest, I have not tested the paging at all.

### 3. I'd love to see the owners' names for the translations

This goes back to the introductory comment; the API doesn't do any creation of the content returned; it just returns whatever it receives from the underlying code. When I look at what is returned from a single sentence https://dev.tatoeba.org/en/api_v0/sentence/1011286 I would expect to see the owner, but in fact it isn't there at all. The owner is definitely an attribute for a sentence, and so this is definitely a place for improvement. Nonetheless, I don't want to block the API going into production for this nice-to-have.

**Comment by ckjpn** [comment]
No problem. Doing whatever is the easiest if sometimes the best thing to do.

The only thing I've ever programmed using CakePHP for Tatoeba.org is this light-weight page.
* show the main sentence (clicakable)
* show only directly-linked translations
I only looked into the code enough to figure out how to do this and not much further.

Here's an example.
https://tatoeba.org/s/s/953070
This is the URL to the regular page.
https://tatoeba.org/en/sentences/show/953070

If I were going to tweek this, I'd probably add in the audio player for the main sentence, if the audio existed.

**Comment by trang** [comment]
@ckjpn 

> Looking at the last line of the output from my "Boston" search above, using "jason-pretty print" I noticed that the output also includes lists that owners have chosen to have "unlisted" (or perhaps "private").

Thanks for pointing this out. I've reworked the code so that such lists are not shown anymore. Note that the lists will only show up if you open the API URL in your browser while having a session as authenticated user. If you try to open the API URL while in incognito mode, or while being logged out, or if you don't use a browser, you shouldn't see lists at all.

@githubshrek 

I've changed slightly the output of the search to include paging information and to restrict the results to 10 results per page. I decided to make this change because I was a bit concerned that returning 100 results could lead to performance issues. And having this limit of 10 results per page also matches the behavior that we have on the website, in the sense that non authenticated users only get 10 results per page and authenticated users also get 10 results per page by default (but they can increase it to 100 results per page in their Settings).

To navigate from a page to another, you simply add the query parameter `page=X`. For example:
https://dev.tatoeba.org/eng/api_v0/search?from=eng&list=907&query=Canadian&trans_filter=limit&trans_link=direct&trans_to=jpn&to=jpn&page=3

If you (or other people) have a use case where you'd need more than 10 results per page, we could consider introducing later on a parameter to specify the desired number of results per page.

**Comment by ckjpn** [comment]
> To navigate from a page to another, you simply add the query parameter page=X. For example:
https://dev.tatoeba.org/eng/api_v0/search?from=eng&list=907&query=Canadian&trans_filter=limit&trans_link=direct&trans_to=jpn&to=jpn&page=3

Is there any easy way to know how many pages of results there are, assuming one has a reason to want to get them all?

**Comment by trang** [comment]
Okay, `page=3` was a bad example, I should have used `page=2` for that URL :)

You can find out how many pages there thanks to the `count` and the `perPage` in the `paging` information.

https://dev.tatoeba.org/eng/api_v0/search?from=eng&list=907&query=Canadian&trans_filter=limit&trans_link=direct&trans_to=jpn&to=jpn&page=2
```
  "paging": {
    "Sentences": {
      "finder": "all",
      "page": 2,
      "current": 8,
      "count": 18,
      "perPage": 10,
      "start": 11,
      "end": 18,
      "prevPage": true,
      "nextPage": false,
      "pageCount": 2,
      "sort": null,
      "direction": null,
      "limit": null,
      "sortDefault": false,
      "directionDefault": false,
      "scope": null,
      "completeSort": []
    }
  },
```

The properties you'll be interested in are:
* `page` - this is the current page.
* `current` - this is the number of items displayed in the current page.
* `count` - this is the total number of results.
* `perPage` - this is the number of results per page.

To know how many pages there are in total, you divide `count` by `perPage` and round up to the upper integer.

**Comment by githubshrek** [comment]
All sound great changes. Thanks @trang

**Comment by trang** [comment]
This is now available on production 🙂

**Comment by ckjpn** [comment]
I started a wiki page on this to briefly summarize the ussage

https://en.wiki.tatoeba.org/articles/show/api

**Comment by ckjpn** [comment]
Question: Does this actually work when accessed from another website?

I can use the URL to download a file and then parse it, but I can't seem to use the URL to directly access the same data from tatoeba.org and parse it.

I wonder if tatoeba.org needs a setting change to allow this.

I'm new to JSON, so maybe it's just because I'm doing something wrong. If so, perhaps someone could suggest what I've done wrong.

## Here are my experiments

### These work, when parsing a pre-downloaded file.

http://tatoeba.ueuo.com/getjason/
http://tatoeba.ueuo.com/getjason/jpn-eng.html
http://tatoeba.ueuo.com/getjason/eng-spa.html

### When accessing the URL, I get "... has been blocked by CORS policy"

http://tatoeba.ueuo.com/getjason/2.html

#### The error message
1 Access to XMLHttpRequest at 'https://tatoeba.org/en/api_v0/search?trans_filter=limit&trans_link=direct&from=jpn&to=eng&has_audio=yes&trans_to=eng&sort=random' from origin 'http://tatoeba.ueuo.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

jquery.min.js:

4 GET https://tatoeba.org/en/api_v0/search?trans_filter=limit&trans_link=direct&from=jpn&to=eng&has_audio=yes&trans_to=eng&sort=random net::ERR_FAILED
send @ jquery.min.js:4
ajax @ jquery.min.js:4
n.<computed> @ jquery.min.js:4
getJSON @ jquery.min.js:4
(anonymous) @ 2.html:53
dispatch @ jquery.min.js:3
r.handle @ jquery.min.js:3

**Comment by ckjpn** [comment]
Does this work?

If so, then perhaps this part of the FAQ needs to be rewritten.

https://en.wiki.tatoeba.org/articles/show/faq#does-tatoeba-provide-an-api?


Does Tatoeba provide an API?
No, it does not (yet).

We unfortunately do not have the proper infrastructure to host a public API. Nonetheless, please do not hesitate to contact us to let us know that you would be interested in this.

With more and more people asking us, we will eventually start something and we will be happy to hear more details about the needs of your application/project.

Meanwhile, what you can do is download our sentences from the Downloads page, then build your own API from there.

**Comment by jiru** [comment]
> Question: Does this actually work when accessed from another website?
> 
> I can use the URL to download a file and then parse it, but I can't seem to use the URL to directly access the same data from tatoeba.org and parse it.

@ckjpn I had a look at this. It does not work indeed. To allow Javascript code executed from a different domain to access the API endpoint, Tatoeba should set the Access-Control-Allow-Origin HTTP header properly. I suggest you open a new issue about that.
